Incorporated PEP 696 into the typing spec. (#1655)
python/typing: docs/spec/generics.rst

   class Bar(Generic[T, ListDefaultT]):
       def __init__(self, x: T, y: ListDefaultT): ...

   reveal_type(Bar)                    # type is type[Bar[T, ListDefaultT = list[T]]]
   reveal_type(Bar[int])               # type is type[Bar[int, list[int]]]
   reveal_type(Bar[int]())             # type is Bar[int, list[int]]
   reveal_type(Bar[int, list[str]]())  # type is Bar[int, list[str]]
   reveal_type(Bar[int, str]())        # type is Bar[int, str]

Specialization Rules
~~~~~~~~~~~~~~~~~~~~

Generic Type Aliases
''''''''''''''''''''

A generic type alias can be further subscripted following normal subscription
rules. If a type parameter has a default that hasn't been overridden, it should
be treated like it was substituted into the type alias.

::

   class SomethingWithNoDefaults(Generic[T, T2]): ...

   MyAlias: TypeAlias = SomethingWithNoDefaults[int, DefaultStrT]  # Valid
   reveal_type(MyAlias)          # type is type[SomethingWithNoDefaults[int, DefaultStrT]]
   reveal_type(MyAlias[bool]())  # type is SomethingWithNoDefaults[int, bool]

   MyAlias[bool, int]  # Invalid: too many arguments passed to MyAlias

Subclassing
'''''''''''

> Generic classes with type parameters that have defaults behave similarly

Reviewer: ```suggestion
Generic classes with type parameters that have defaults behave similarly to
```

   reveal_type(Foo)                  # type is type[Foo[DefaultP = [str, int]]]
   reveal_type(Foo())                # type is Foo[[str, int]]
   reveal_type(Foo[[bool, bool]]())  # type is Foo[[bool, bool]]

``TypeVarTuple`` Defaults
'''''''''''''''''''''''''

``TypeVarTuple`` defaults are defined using the same syntax as
``TypeVar`` \ s but use an unpacked tuple of types instead of a single type
or another in-scope ``TypeVarTuple`` (see `Scoping Rules`_).

::

   DefaultTs = TypeVarTuple("DefaultTs", default=Unpack[tuple[str, int]])

   class Foo(Generic[*DefaultTs]): ...

   reveal_type(Foo)               # type is type[Foo[DefaultTs = *tuple[str, int]]]
   reveal_type(Foo())             # type is Foo[str, int]
   reveal_type(Foo[int, bool]())  # type is Foo[int, bool]

Using Another Type Parameter as ``default``
''''''''''''''''''''''''''''''''''''''''''''

This allows for a value to be used again when the type parameter to a
generic is missing but another type parameter is specified.

To use another type parameter as a default the ``default`` and the
type parameter must be the same type (a ``TypeVar``'s default must be
a ``TypeVar``, etc.).

> `This could be used on builtins.slice <https://github.com/python/typing/issues/159>`__
> where the ``start`` parameter should default to ``int``, ``stop``

Reviewer 2: ```suggestion
For instance, this could be used to type something like builtins.slice,
where the ``start`` parameter should default to ``int``, ``stop``
```
nit: original language was more suited for proposal, sounds dated if typeshed starts using 696 on slice

solution. However, it's the best we can do without additional type
manipulation mechanisms.)

.. _`type_parameter_defaults`:

Defaults for Type Parameters
----------------------------

(Originally specified in :pep:`696`.)

Default values can be provided for a TypeVar, ParamSpec, or TypeVarTuple.

Default Ordering and Subscription Rules
'''''''''''''''''''''''''''''''''''''''

The order for defaults should follow the standard function parameter
rules, so a type parameter with no ``default`` cannot follow one with

> a ``default`` value. Doing so should ideally raise a ``TypeError`` in
> ``typing._GenericAlias``/``types.GenericAlias``, and a type checker

Reviewer: ```suggestion
a ``default`` value. Doing so may raise a ``TypeError`` at runtime, and a type checker
```
It's not the typing spec's job to say what the runtime should do.

Author: Good catch. I caught (and removed) one such statement, but I missed this one.

   TypeVar("Invalid", float, str, default=T1)         # Invalid: upper bound int is incompatible with constraints float or str

   T1 = TypeVar("T1", int, str)
   TypeVar("AlsoOk", int, str, bool, default=T1)      # Valid
   TypeVar("AlsoInvalid", bool, complex, default=T1)  # Invalid: {bool, complex} is not a superset of {int, str}

Type Parameters as Parameters to Generics
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Type parameters are valid as parameters to generics inside of a
``default`` when the first parameter is in scope as determined by the
`previous section <scoping rules_>`_.

::

   T = TypeVar("T")
   ListDefaultT = TypeVar("ListDefaultT", default=list[T])

   class Bar(Generic[T, ListDefaultT]):
       def __init__(self, x: T, y: ListDefaultT): ...

   reveal_type(Bar)                    # type is type[Bar[T, ListDefaultT = list[T]]]
   reveal_type(Bar[int])               # type is type[Bar[int, list[int]]]
   reveal_type(Bar[int]())             # type is Bar[int, list[int]]
   reveal_type(Bar[int, list[str]]())  # type is Bar[int, list[str]]
   reveal_type(Bar[int, str]())        # type is Bar[int, str]

Specialization Rules
~~~~~~~~~~~~~~~~~~~~

> ``Generic`` ``TypeAlias``\ es
> '''''''''''''''''''''''''''''
> 
> ``Generic`` ``TypeAlias``\ es can be further subscripted

Reviewer: ```suggestion
Generic type aliases
'''''''''''''''''''''''''

Generic type aliases can be further subscripted
```
"Generic" here refers to the concept of a generic, not the class `typing.Generic`, so it should not be monospaced. Similarly, this section should apply to all type aliases, whether or not they're created with `typing.TypeAlias`, so "TypeAlias" should also not be monospaced.

following normal subscription rules. If a type parameter has a default
that hasn't been overridden, it should be treated like it was

> substituted into the ``TypeAlias``. However, it can be specialized

Reviewer: ```suggestion
substituted into the type alias. However, it can be specialized
```

further down the line.

::

   class SomethingWithNoDefaults(Generic[T, T2]): ...

   MyAlias: TypeAlias = SomethingWithNoDefaults[int, DefaultStrT]  # Valid
   reveal_type(MyAlias)          # type is type[SomethingWithNoDefaults[int, DefaultStrT]]
   reveal_type(MyAlias[bool]())  # type is SomethingWithNoDefaults[int, bool]

   MyAlias[bool, int]  # Invalid: too many arguments passed to MyAlias

Subclassing
'''''''''''

> Subclasses of ``Generic``\ s with type parameters that have defaults
> behave similarly to ``Generic`` ``TypeAlias``\ es. That is, subclasses can be

Reviewer: ```suggestion
Subclasses of generic classes with type parameters that have defaults
behave similarly to generic type aliases. That is, subclasses can be
```
